### ISSUE TITLE
Agroal Data Source Health check failing for reactive data source

### ISSUE THREAD

**Issue by mvmandapati** [quarkusio/quarkus#38934]
### Describe the bug

When application defines a mix of datasources (Reactive & JDBC), health check fails with below error.

Works fine with Quarkus Version < 3.7. I think issue is introduced by https://github.com/quarkusio/quarkus/pull/38108

Sample config:

quarkus.datasource.db-kind=postgresql
quarkus.datasource.metrics.enabled=true
quarkus.datasource.jdbc=false
quarkus.datasource.reactive=true
## additional datasource for flyway migration
quarkus.datasource.migds.db-kind=postgresql
quarkus.datasource.migds.reactive=false
quarkus.datasource.migds.jdbc=true
quarkus.flyway.migds.clean-at-start=false
quarkus.flyway.migds.create-schemas=false
quarkus.flyway.migds.migrate-at-start=true
quarkus.flyway.migds.repair-at-start=true

### Expected behavior

Health check should return health status of both the datasources.

### Actual behavior

Start app using quarkus:dev

Test URL http://[localhost:8080/q/health](http://localhost:8080/q/health)

```
{
    "status": "DOWN",
    "checks": [
        {
            "name": "io.quarkus.agroal.runtime.health.DataSourceHealthCheck_ClientProxy",
            "status": "DOWN",
            "data": {
                "exceptionClass": "java.lang.IllegalArgumentException",
                "exceptionMessage": "No datasource named '<default>' exists",
                "rootCause": "No datasource named '<default>' exists"
            }
        },
        {
            "name": "Reactive PostgreSQL connections health check",
            "status": "UP",
            "data": {
                "<default>": "UP"
            }
        }
    ]
}
```

Errors in Logs

```
SRHCK01000: Error processing Health Checks: java.lang.IllegalArgumentException: No datasource named '<default>' exists
        at io.quarkus.agroal.runtime.DataSources.doCreateDataSource(DataSources.java:165)
        at io.quarkus.agroal.runtime.DataSources$1.apply(DataSources.java:145)
        at io.quarkus.agroal.runtime.DataSources$1.apply(DataSources.java:142)
        at java.base/java.util.concurrent.ConcurrentHashMap.computeIfAbsent(ConcurrentHashMap.java:1708)
        at io.quarkus.agroal.runtime.DataSources.getDataSource(DataSources.java:142)
        at io.quarkus.agroal.runtime.health.DataSourceHealthCheck.init(DataSourceHealthCheck.java:49)

```

### How to Reproduce?

Easy to reproduce. Getting started sample with below config produces the error.

```
quarkus.datasource.db-kind=postgresql
quarkus.datasource.metrics.enabled=true
quarkus.datasource.jdbc=false
quarkus.datasource.reactive=true
## additional datasource for flyway migration
quarkus.datasource.migds.db-kind=postgresql
quarkus.datasource.migds.reactive=false
quarkus.datasource.migds.jdbc=true
quarkus.flyway.migds.clean-at-start=false
quarkus.flyway.migds.create-schemas=false
quarkus.flyway.migds.migrate-at-start=true
quarkus.flyway.migds.repair-at-start=true
```

Sample app: https://github.com/mvmandapati/code-with-quarkus

### Output of `uname -a` or `ver`

Linux

### Output of `java -version`

openjdk 21.0.2 2024-01-16

### Quarkus version or git rev

3.7.3

### Build tool (ie. output of `mvnw --version` or `gradlew --version`)

Maven 3.9.1

**Comment by gsmet** [comment]
Outch, there are multiple issues here as far as I can see. I will have a look.

**Comment by gsmet** [comment]
Thanks for the report. https://github.com/quarkusio/quarkus/pull/38959 should fix it.

**Comment by mvmandapati** [comment]
Thanks for looking into the issue. Just want to check if the fix will be backported to 3.7.x or 3.8?

**Comment by gsmet** [comment]
Yes it's planned for being backported to 3.8.

**Comment by Shweta-Shah-BP** [comment]
Hey, thanks for fixing this!
Is the fix planned to be backported to 2.16.12.Final?
or is there a recommended workaround to make it work the lesser versions?

**Comment by gsmet** [comment]
2.16 is not supported anymore so there won't be fixes there. But the issue explicitly states that the problem started occurring with Quarkus 3.7+ so I'm not even sure you would be affected by this specific issue (might be another one).

I would recommend to move to 3.8 as you won't get any security fixes in 2.16.

You can use `quarkus update` to do most of the grunt work of updating to 3.8.
